File: vault-audit-tools/src/utils/time.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};

/// Parse a timestamp string from Vault audit logs
#[allow(dead_code)]
pub fn parse_timestamp(ts: &str) -> Result<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(ts)
        .context("Failed to parse timestamp")
        .map(|dt| dt.with_timezone(&Utc))
}

/// Format a timestamp for display
#[allow(dead_code)]
pub fn format_timestamp(dt: &DateTime<Utc>) -> String {
    dt.format("%Y-%m-%d %H:%M:%S UTC").to_string()
}

/// Calculate duration between two timestamps in human-readable format
#[allow(dead_code)]
pub fn duration_human(start: &DateTime<Utc>, end: &DateTime<Utc>) -> String {
    let duration = end.signed_duration_since(*start);
    let seconds = duration.num_seconds();

    if seconds < 60 {
        format!("{} seconds", seconds)
    } else if seconds < 3600 {
        format!("{} minutes", seconds / 60)
    } else if seconds < 86400 {
        format!("{:.1} hours", seconds as f64 / 3600.0)
    } else {
        format!("{:.1} days", seconds as f64 / 86400.0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Datelike;

    #[test]
    fn test_parse_timestamp() {
        let ts = "2025-10-06T07:26:03.801191678Z";
        let dt = parse_timestamp(ts).unwrap();
        assert_eq!(dt.year(), 2025);
        assert_eq!(dt.month(), 10);
        assert_eq!(dt.day(), 6);
    }

    #[test]
    fn test_duration_human() {
        let start = parse_timestamp("2025-10-06T07:26:03Z").unwrap();
        let end = parse_timestamp("2025-10-06T08:26:03Z").unwrap();
        let duration = duration_human(&start, &end);
        assert!(duration.contains("1.0 hours"));
    }
}

File: internal/utils/time.go
package utils

import (
	"fmt"
	"time"
)

// ParseTimestamp parses an RFC 3339 / ISO 8601 timestamp string as used
// in Vault audit logs (e.g. "2025-10-07T12:00:00.000Z").
func ParseTimestamp(ts string) (time.Time, error) {
	t, err := time.Parse(time.RFC3339Nano, ts)
	if err != nil {
		// Fallback: try without nanoseconds
		t, err = time.Parse(time.RFC3339, ts)
	}
	return t, err
}

// FormatTimestamp formats a time.Time for human-readable display.
func FormatTimestamp(t time.Time) string {
	return t.UTC().Format("2006-01-02 15:04:05 UTC")
}

// DurationHuman returns a human-readable string for the duration between
// two timestamps (e.g. "3.5 hours", "2.1 days").
func DurationHuman(start, end time.Time) string {
	d := end.Sub(start)
	seconds := int64(d.Seconds())

	switch {
	case seconds < 60:
		return fmt.Sprintf("%d seconds", seconds)
	case seconds < 3600:
		return fmt.Sprintf("%d minutes", seconds/60)
	case seconds < 86400:
		return fmt.Sprintf("%.1f hours", float64(seconds)/3600)
	default:
		return fmt.Sprintf("%.1f days", float64(seconds)/86400)
	}
}

// HoursBetween returns the number of hours between two RFC 3339 timestamp strings.
// Returns 0 and a non-nil error if either string cannot be parsed.
func HoursBetween(first, last string) (float64, error) {
	t1, err := ParseTimestamp(first)
	if err != nil {
		return 0, fmt.Errorf("parse first timestamp %q: %w", first, err)
	}
	t2, err := ParseTimestamp(last)
	if err != nil {
		return 0, fmt.Errorf("parse last timestamp %q: %w", last, err)
	}
	return t2.Sub(t1).Hours(), nil
}
